services/webtheme/webtheme.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package webtheme
+
+import (
+	"strings"
+	"sync"
+
+	"code.gitea.io/gitea/modules/container"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/public"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+var (
+	availableThemes    []string
+	availableThemesSet container.Set[string]
+	themeOnce          sync.Once
+)
+
+func initThemes() {
+	availableThemes = nil
+	defer func() {
+		availableThemesSet = container.SetOf(availableThemes...)
+	}()
+	cssFiles, err := public.AssetFS().ListFiles("/assets/css")
+	if err != nil {
+		log.Error("Failed to list themes: %v", err)
+		availableThemes = []string{setting.UI.DefaultTheme}
+		return
+	}
+	var foundThemes []string
+	for _, name := range cssFiles {
+		name, ok := strings.CutPrefix(name, "theme-")
+		if !ok {
+			continue
+		}
+		name, ok = strings.CutSuffix(name, ".css")
+		if !ok {
+			continue
+		}
+		foundThemes = append(foundThemes, name)
+	}
+	if len(setting.UI.Themes) > 0 {
+		allowedThemes := container.SetOf(setting.UI.Themes...)
+		for _, theme := range foundThemes {
+			if allowedThemes.Contains(theme) {
+				availableThemes = append(availableThemes, theme)
+			}
+		}
+	} else {
+		availableThemes = foundThemes
+	}
+	if len(availableThemes) == 0 {
+		log.Error("No theme candidate, but gitea requires there should be at least one usable theme")
+		availableThemes = []string{setting.UI.DefaultTheme}
+	}
+}
+
+func GetAvailableThemes() []string {
+	themeOnce.Do(initThemes)
```

Comment:
If files are changed, user may get a broken UI. maybe it is acceptable.
But can we improve the settings page here?
![image](https://github.com/go-gitea/gitea/assets/18380374/a30c23a1-9492-4702-b9bf-7c5bc4c8c2ac)

Reply:
Cool, you pointed out an old bug, without this PR, the users might see broken UI if the themes are changed. This bug has been fixed by the new `UserThemeName`.

However, I decided to leave "theme" as-is because I think such UI could make users clearly know that "themes have changed" and lead them to choose a new theme manually.

Reply:
Maybe some notifications here is better. Or maybe you will see some issues about `why my theme is empty`
![image](https://github.com/go-gitea/gitea/assets/18380374/63005a8e-62b3-423c-abb5-48bdeed8190b)

Reply:
To avoid the "empty" theme, it could be like this: 36347f364197717b00e8aaecd76b512dd9a16b0b

(related code is also simplified)